### PR TITLE
Clone objects

### PR DESCRIPTION
**Pull request by hishnash**

### CODE REVIEW

File: src/gitbackend.c [hunk 1]
```diff
@@ -345,3 +345,70 @@ int init_repo(git_repository** repo, char* path, Archive* archive) {
     }
     return attach_archive_to_repo(*repo, archive);
 }
+
+#include <Python.h>
+
+typedef struct CRepository {
```
Comment (hishnash) [inline src/gitbackend.c]: this is a private header so i recreate the struct here so i can get at the repo pointer

File: src/gitbackend.c [hunk 2]
```diff
@@ -345,3 +345,70 @@ int init_repo(git_repository** repo, char* path, Archive* archive) {
...
+    PyObject_HEAD
+    git_repository* repo;
+    PyObject *index;  /* It will be None for a bare repository */
+    PyObject *config; /* It will be None for a bare repository */
+    int owned;    /* _from_c() sometimes means we don't own the C pointer */
+} CRepository;
+
+
+int clone_object(PyObject* source_repo_object, PyObject* target_repo_object, const char* hex) {
```
Comment (hishnash) [inline src/gitbackend.c]: PyObjects are structs that extend the `PyObject_HEAD`

File: src/gitbackend.c [hunk 3]
```diff
@@ -345,3 +345,70 @@ int init_repo(git_repository** repo, char* path, Archive* archive) {
...
+    if (memcmp(new_oid.id, oid.id, 20 * sizeof(char)) != 0) {
+        return -1;
+    }
+
+
+    if (git_odb_exists(target_odb, &oid) != 1) {
+        return -1;
+    }
+
+    git_odb_object_free(object);
```
Comment (hishnash) [inline src/gitbackend.c]: this will free the `data` as well.

File: src/gitbackend.c (hunk 1 above)
Comment (mittsh) [inline src/gitbackend.c]: ok

File: src/gitbackend.c (hunk 2 above)
Comment (mittsh) [inline src/gitbackend.c]: ok

File: src/gitbackend.c [hunk 4]
```diff
@@ -345,3 +345,70 @@ int init_repo(git_repository** repo, char* path, Archive* archive) {
...
+    int owned;    /* _from_c() sometimes means we don't own the C pointer */
+} CRepository;
+
+
+int clone_object(PyObject* source_repo_object, PyObject* target_repo_object, const char* hex) {
+    git_odb* source_odb;
+    git_odb* target_odb;
+    git_repository* source_repo = ((CRepository*) source_repo_object)->repo;
+    git_repository* target_repo = ((CRepository*) target_repo_object)->repo;
+    int error = git_repository_odb(&source_odb, source_repo);
```
Comment (mittsh) [inline src/gitbackend.c]: Are you calling this for every single object? Isn't this a performance hit vs. just instantiating this once for all?

File: src/gitbackend.c [hunk 5]
```diff
@@ -345,3 +345,70 @@ int init_repo(git_repository** repo, char* path, Archive* archive) {
...
+    git_otype type = git_odb_object_type(object);
+
+    git_oid new_oid;
+
+    error = git_odb_write(&new_oid, target_odb, data, size, type);
+    if (error < 0) {
+        return error;
+    }
+
+    if (memcmp(new_oid.id, oid.id, 20 * sizeof(char)) != 0) {
```
Comment (mittsh) [inline src/gitbackend.c]: How about: `memcmp(new_oid.id, oid.id, sizeof(new_oid.id))` or if it fails: `memcmp(new_oid.id, oid.id, GIT_OID_RAWSZ * sizeof(unsigned char))` so we conform with the `oid.h` header?

**Review by mittsh — APPROVED** [review]
Looks good! So to my understanding this only implements the migration of 1 object at a time? Then Python will iterate over all objects?

I've left a few minor comments, nothing important.

File: src/gitbackend.c (hunk 4 above)
Comment (hishnash) [inline src/gitbackend.c]: yes but we do this once per project, this way I keep it nicely scopes and will get a clear python takctrack on the object that fialed if someting fails

File: src/gitbackend.c (hunk 5 above)
Comment (hishnash) [inline src/gitbackend.c]: i wanted to be explicit since we are explicit in our archive code.

Comment (mittsh) [inline src/gitbackend.c]: ok!

File: src/gitbackend.c (hunk 4 above)
Comment (mittsh) [inline src/gitbackend.c]: sure :)

**Comment by mittsh** [comment]
Then all looks good @hishnash merge at your own pace

**Pull request merged**